### PR TITLE
feat(diff): detect splits and merges on the content-matching residue (#1891)

### PR DESCRIPTION
**Pull request by louistrue**
The last unbuilt piece of #1891, and the thing @bruadam specifically asked for. Two commits: volume plumbing, then the detector.

## Real-fixture evidence

`AC20-FZK-Haus.ifc`, interior wall #18465 (3.71 x 0.24 x 2.5 m) cut into three 1.23 m segments with a **10 mm joint** between each pair, then **every IfcRoot re-GUIDed**. Driven through the real wasm pass with hashing on:

| | |
|---|---|
| whole, proved volume | **2.226000 m3** (closure `0x0F`) |
| each piece | **0.738000 m3** x3 |
| pieces sum | **2.214000 m3** |
| residual | **-0.539 %** (the two joints) |
| result | **1 claim** - `split` / `verified` |
| sides swapped | **1 claim** - `merge`, identical numbers |

**The negative controls matter more than the positive one.** A detector that fires on a pure re-GUID is worse than no detector:

| model | residue | claims |
|---|---|---|
| AC20-FZK-Haus | 111 / 111 (+4 ambiguous groups) | **0** |
| duplex.ifc | 277 / 277 | **0** |
| IfcOpenHouse_IFC4 | 5 / 5 | **0** |
| C20-Institute-Var-2 | 903 / 903 | **0** |

Volume coverage on these files: **65-66%** of hashed entities.

## Your four decisions, as built

3% relative tolerance, gated. Padding `max(0.05 m, 1% of diagonal)`. Split-and-moved reported at lower confidence. No semantic piece limit (256 is a perf bail only).

**The two interpretations I flagged are implemented as written and documented as interpretations in the commit body**: `displaced` requires complete trusted volume (decision 3 forbids *policy* refusal, not a claim where evidence cannot exist), and the padding is container-side (for merge the container is the *added* element).

## Three places I went beyond the brief

**1. Federation alignment withholds the volume.** Alignment applies a **scale**, so a pre-alignment volume describes geometry no longer on screen. The box survives because it is re-measured; a volume cannot be, since the closure proof lives in the kernel. Rather than scale by a determinant the cross-CRS path does not have, the volume is withheld on exactly `same-crs`/`reprojected`.

**2. Lane 2 vetoes a cluster still inside its whole.** Without it, a sub-cluster of a containment set that lane 1 had **refuted** could return wearing a `displaced` label - subset search by the back door, over pieces that never moved. Reachable, and pinned.

**3. `types.ts` was split** - already 427 lines, and these additions would have taken it to 616. Now 480. **Still over the guideline**, flagged rather than glossed.

## Verification

44 mutations, 44 killed. `@ifc-lite/diff` 261 - `@ifc-lite/geometry` 187 - viewer 2341 - typecheck 34/34 - `pnpm test` 74/74 - wasm contract 43/43 - api-surface clean.

**Three fixtures were rewritten after a mutation proved the test could not fail.** This is the recurring class and worth reading:

- `wholeWall(undefined)` - a JS **default parameter fires on `undefined`**, so the "whole has no volume" test silently handed back the *measured* wall and asserted the opposite of its own name.
- The lane-1 containment decoy sat 100 m away, where **the spatial grid never offers it** - the containment predicate was never reached, so the test passed with or without it.
- The `maxSplitPieces` fixture would have failed the volume gate after truncation anyway, so truncation was not what it tested.

**One guard was found genuinely unfalsifiable and its test deleted rather than pinned**: a piece volume is positive by construction, so it can never be within tolerance of a *negative* overshoot. The within-tolerance half of the same guard is falsifiable and now has a test that bites.

Also caught post-commit: the file contained a literal NUL byte as its tiebreak separator, which made git treat it as binary. Now written as an escape.

## What it cannot see, documented

Cross-class splits (candidates are per-`ifcType`); two or more same-class interlopers, since the overshoot is their *sum* and no single piece explains it; `extent` fires on redesign-in-place and on a perimeter enclosing an unfilled middle; `displaced` abstains between two congruent clusters in a repetitive building; non-90 degree rotated moved splits; and **a >3% material change with full volume data is refused** - decision 1's direct cost.

Non-destructive throughout: it never retires entries and never touches `counts`. A split claim binds k+1 entities on one evidence chain, and retiring them would let one wrong claim delete k+1 real changes.

Toward #1891.

## Summary by CodeRabbit

* **New Features**
  * Added optional proved enclosed-volume data to geometry results and model fingerprints.
  * Added opt-in split and merge detection for model comparisons, with confidence levels, volume evidence, displacement analysis, and conflict handling.
  * Added public configuration and result types for split/merge claims.
* **Documentation**
  * Updated API and model comparison guides with volume handling, detection options, limitations, and examples.
* **Tests**
  * Expanded coverage for geometry volumes, alignment behavior, split/merge detection, conflict resolution, and edge cases.

### CODE REVIEW

File: packages/diff/src/split-merge.ts [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+      continue;
+    }
+
+    // Indistinguishable evidence: the only thing that put the incumbent first
+    // was the alphabet. Retire it too rather than let the alphabet decide.
+    for (const blocker of blockers) {
+      if (compareRank(blocker.rank, candidate.rank) !== 0) continue;
+      accepted.delete(blocker);
+      for (const entity of participants(blocker.claim)) claimedBy.delete(entity);
+    }
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/diff/src/split-merge.ts:269]: Correct, and the diagnosis was exactly right: the abstention was implemented pairwise, and pairwise does not generalise. The second claim retired the first, freed everything it held, and the third walked into the gap.

Fixed by **remembering the tie** rather than re-pairing the leftover. Re-pairing is the shape that works at three claims and fails at four; a persistent contested-set makes the arity irrelevant by construction. Participants of both sides are recorded as contested at that rank, and a later claim of exactly that rank touching any of them is refused. Tested at three and at four (four claims all binding `a` gives `[]`), plus a cycle `{a,b} {b,c} {a,c}` and a multi-blocker case.

Two scoping decisions inside it, and both only got tests because a mutation survived:

- The gap is the **whole participant set** of both sides, not just the entity they collided on, because retiring the incumbent frees everything it held.
- The **loser's** material is contested too, even though nothing of its was ever freed. Handing it to the next equal claim would decide "the loser loses this" on the strength of it having lost a tie about something else, which is the alphabet one hop removed. This was the one mutation that survived the first round; the code turned out to be right, so it got a discriminating test (`{a,b} / {a,c} / {c,e}` all equal gives `[]`) rather than a deleted line.

Rank is stored rather than a flag, so the documented "a strictly weaker claim still stands" rule survives, and the flag-instead-of-rank mutation is killed by that pin.

File: packages/diff/src/split-merge-geometry.ts [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+      walk = next;
+    }
+    return root;
+  };
+
+  const grid = buildBoxGrid(boxes);
+  for (let i = 0; i < count; i++) {
+    for (const j of grid.query(boxes[i], paddings[i])) {
+      if (j <= i) continue;
+      if (!boxesTouch(boxes[i], boxes[j], Math.max(paddings[i], paddings[j]))) continue;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/diff/src/split-merge-geometry.ts:274]: Fixed. The query now reaches `2 * max(padding)`. Worth noting for anyone reading later: `pad_i + maxPad` does **not** dominate `2 * max(pad_i, pad_j)` when the two differ, so the global doubling is the correct bound rather than the tempting per-box one.

The more useful part is what happened to the fixture. I built a wide-joint case on real geometry (same wall, two 1.815 m segments, an 80 mm joint against the 50 mm padding floor, shifted 30 m along its own axis). It reports split/displaced, 2.178001 m3 against 2.226000 m3, residual -2.156 %. That is good end-to-end evidence that the two-width path works.

**It is not a regression test for this defect**, and I checked rather than assumed: reverting the query reach on that fixture still produces the claim. The cell size is the largest box in the residue, 18.000 m on this model, and a cell that size swallows an 80 mm gap whole. The alignment simply happened to be favourable, which is this finding restated.

So both are committed and the commit body says which is which. The actual regression test is the unit case, where cell size is controlled at 1.0 m and the boundary sits inside the gap so box 0 is in cell 0 alone, box 1 in cells 1 and 2, and the reverse query is discarded by `j <= i`.

File: packages/diff/src/types.ts [hunk 3]
```diff
@@ -316,79 +418,20 @@ export interface DiffCounts {
...
-  | 'reshaped'
-  | 'duplicated'
-  | 'deduplicated'
-  | 'ambiguous';
+export type { ContentMatch, ContentMatchKind } from './content-match-types.js';
+export type {
+  SplitMergeClaim,
+  SplitMergeConfidence,
+  SplitMergeKind,
+} from './split-merge-types.js';
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/diff/src/types.ts:191]: Fixed. `DiffOptions` moved to `diff-options.ts` (257 lines), re-exported alongside the two record modules. `types.ts` is now 240. All four are under the guideline: 240 / 257 / 89 / 96.

No consumer import changed; `./types.js` is still the single import site. The api-surface snapshot is byte-identical (37 packages, 58 export surfaces, 3988 exports, no update needed), which is the check that this was a pure move and not a quiet surface change.

**Comment by louistrue** [comment]
@coderabbitai full review

File: packages/diff/src/split-merge-lanes.ts [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+  splitPaddingRatio?: number;
+  maxSplitPieces?: number;
+}): SplitMergeSettings {
+  return {
+    volumeTolerance: positiveOr(options.splitVolumeTolerance, DEFAULT_SPLIT_VOLUME_TOLERANCE),
+    paddingMin: positiveOr(options.splitPaddingMin, DEFAULT_SPLIT_PADDING_MIN),
+    paddingRatio: positiveOr(options.splitPaddingRatio, DEFAULT_SPLIT_PADDING_RATIO),
+    maxPieces: positiveOr(options.maxSplitPieces, DEFAULT_MAX_SPLIT_PIECES),
+  };
+}
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/diff/src/split-merge-lanes.ts]: Fixed, and the sting is fair: I wrote the comment that says the coercion exists so an untyped caller cannot make every claim silently vanish, and `maxSplitPieces: 1` does exactly that through a different door. `positiveOr` accepts any finite value `>= 0`, so `0`, `1` and `2.5` all passed, and `0` or `1` then rejects every candidate both lanes can build, since a candidate has at least `MIN_PIECES` pieces by definition.

The cap now goes through its own `pieceCapOr`: finite, floored to an integer, at least `MIN_PIECES`, else the documented default.

**Fall back to the default rather than clamp or reject**, in increasing order of weight:

1. Not reject: `diffModels` has no error channel for options and every other one coerces. Throwing would be a new failure mode in a package whose stated contract here is that an untyped caller cannot poison a run, and a behavioural break for anyone passing junk today.
2. Not clamp: clamping `1` to `2` invents an intent the caller never expressed. There is no split into fewer than two pieces, so `1` is not a tighter cap, it is an unachievable one. Clamping still overrides silently and produces a run behaving like neither `1` nor the default.
3. The decisive and knob-specific one: `maxSplitPieces` is documented as a performance bail, never a semantic rule. A cap resolving **higher** than asked can only cost time on a pathological model; it can never produce a wrong claim. A cap resolving to nothing produces a wrong answer. For this knob, more-permissive is strictly the safer direction.

That third argument is also why this is deliberately **not** a widened shared helper: `pieceCapOr` applied to a tolerance would floor `0.03` to `0`, and a zero tolerance refuses every claim while a zero padding refuses every containment. That is the same silent detect-nothing moved one field left, so there is a mutation that does precisely that and a test that kills it.

Fractional values are floored rather than refused when still achievable, because a cap of 2.5 pieces is a cap of 2 pieces and counts are integers. `1.9` floors to `1`, is unachievable, and falls back.

Resolution: `0 / 1 / 1.9 / -4 / NaN / +-Infinity / undefined` all give 256; `2` gives 2; `2.5` gives 2; `9.99` gives 9.

End-to-end: `maxSplitPieces: 0 | 1 | 1.9 | -4` each now produce the 3-piece `verified` claim where previously they produced nothing, silently. `2.5` correctly drops a 3-piece candidate and keeps a 2-piece one, so the cap is applied rather than ignored.

Both alternatives I rejected are pinned as mutations rather than only argued here, so the decision cannot be quietly reversed later.

File: packages/diff/src/diff-options.ts [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+  splitPaddingRatio?: number;
+  /**
+   * Largest number of pieces one claim may bind. Default `256`.
+   *
+   * A PERFORMANCE bail, never a semantic rule: a precast slab field really is
+   * dozens of panels, and there is no piece count at which a split stops being
+   * a split. Above the cap the candidate is dropped rather than truncated —
+   * reporting the first 256 of 400 panels would be a claim nobody made.
+   */
+  maxSplitPieces?: number;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/diff/src/diff-options.ts:269]: Fixed, and the sting is fair: I wrote the comment that says the coercion exists so an untyped caller cannot make every claim silently vanish, and `maxSplitPieces: 1` does exactly that through a different door. `positiveOr` accepts any finite value `>= 0`, so `0`, `1` and `2.5` all passed, and `0` or `1` then rejects every candidate both lanes can build, since a candidate has at least `MIN_PIECES` pieces by definition.

The cap now goes through its own `pieceCapOr`: finite, floored to an integer, at least `MIN_PIECES`, else the documented default.

**Fall back to the default rather than clamp or reject**, in increasing order of weight:

1. Not reject: `diffModels` has no error channel for options and every other one coerces. Throwing would be a new failure mode in a package whose stated contract here is that an untyped caller cannot poison a run, and a behavioural break for anyone passing junk today.
2. Not clamp: clamping `1` to `2` invents an intent the caller never expressed. There is no split into fewer than two pieces, so `1` is not a tighter cap, it is an unachievable one. Clamping still overrides silently and produces a run behaving like neither `1` nor the default.
3. The decisive and knob-specific one: `maxSplitPieces` is documented as a performance bail, never a semantic rule. A cap resolving **higher** than asked can only cost time on a pathological model; it can never produce a wrong claim. A cap resolving to nothing produces a wrong answer. For this knob, more-permissive is strictly the safer direction.

That third argument is also why this is deliberately **not** a widened shared helper: `pieceCapOr` applied to a tolerance would floor `0.03` to `0`, and a zero tolerance refuses every claim while a zero padding refuses every containment. That is the same silent detect-nothing moved one field left, so there is a mutation that does precisely that and a test that kills it.

Fractional values are floored rather than refused when still achievable, because a cap of 2.5 pieces is a cap of 2 pieces and counts are integers. `1.9` floors to `1`, is unachievable, and falls back.

Resolution: `0 / 1 / 1.9 / -4 / NaN / +-Infinity / undefined` all give 256; `2` gives 2; `2.5` gives 2; `9.99` gives 9.

End-to-end: `maxSplitPieces: 0 | 1 | 1.9 | -4` each now produce the 3-piece `verified` claim where previously they produced nothing, silently. `2.5` correctly drops a 3-piece candidate and keeps a 2-piece one, so the cap is applied rather than ignored.

Both alternatives I rejected are pinned as mutations rather than only argued here, so the decision cannot be quietly reversed later.

File: docs/guide/model-diff.md [hunk 6]
```diff
@@ -186,7 +188,7 @@ Ambiguous groups retire nothing, so a collision landing in one costs the caller
 
 The residual concentrates in the **1:1 leftover**. Every other retiring path has corroboration beyond the data hash — an agreeing world geometry hash, or a mutual-nearest-neighbour agreement within the move cap — while the 1:1 leftover rests on the data hash, `ifcType`, and `components` alone. That is where a false pair would come from.
 
-Split and Merged (a *partial* geometric overlap between one entity and several others) are not implemented — they need a geometric-similarity threshold and a partial-overlap policy with no single correct answer.
+Split and Merged — a *partial* geometric overlap between one entity and several others — are a separate, purely additive stage on this pass's residue: see [Split and merge detection](#split-and-merge-detection). They are not `ContentMatch`es and retire nothing.
```
[the remark this answers is not in the document]
Comment (louistrue) [reply docs/guide/model-diff.md]: Fixed in both places the sentence lived: `docs/guide/model-diff.md:191` and the `matchUnpairedByContent` doc comment in `diff-options.ts`. Both now read "Splits and merges ... are a separate, purely additive stage", matching the two claim kinds the rest of the document and the `SplitMergeKind` union already use.

I left `packages/diff/CHANGELOG.md` alone deliberately: it carries the old wording, but it is the generated record of what shipped in 0.5.0 rather than prose to maintain.
